src/vs/workbench/contrib/externalTerminal/browser/externalTerminal.contribution.ts

```diff
@@ -26,114 +26,195 @@ import { LifecyclePhase } from 'vs/workbench/services/lifecycle/common/lifecycle
+					const cwds = targets.map(({ stat }) => {
+						const resource = stat!.resource;
+						if (stat!.isDirectory) {
+							return resource;
+						}
+						return URI.from({
+							scheme: resource.scheme,
+							authority: resource.authority,
+							fragment: resource.fragment,
+							query: resource.query,
+							path: dirname(resource.path)
+						});
 					});
-				});
-				for (const cwd of cwds) {
-					if (opened[cwd.path]) {
-						return;
-					}
-					opened[cwd.path] = true;
-					const instance = await integratedTerminalService.createTerminal({ config: { cwd } });
-					if (instance && instance.target !== TerminalLocation.Editor && (resources.length === 1 || !resource || cwd.path === resource.path || cwd.path === dirname(resource.path))) {
-						integratedTerminalService.setActiveInstance(instance);
-						terminalGroupService.showPanel(true);
+					for (const cwd of cwds) {
+						if (opened[cwd.path]) {
+							return;
+						}
+						opened[cwd.path] = true;
+						const instance = await integratedTerminalService.createTerminal({ config: { cwd } });
+						if (instance && instance.target !== TerminalLocation.Editor && (resources.length === 1 || !resource || cwd.path === resource.path || cwd.path === dirname(resource.path))) {
+							integratedTerminalService.setActiveInstance(instance);
+							terminalGroupService.showPanel(true);
+						}
 					}
+				} else if (externalTerminalService) {
+					distinct(targets.map(({ stat }) => stat!.isDirectory ? stat!.resource.fsPath : dirname(stat!.resource.fsPath))).forEach(cwd => {
+						externalTerminalService!.openTerminal(config.terminal.external, cwd);
+					});
 				}
-			} else if (externalTerminalService) {
-				distinct(targets.map(({ stat }) => stat!.isDirectory ? stat!.resource.fsPath : dirname(stat!.resource.fsPath))).forEach(cwd => {
-					externalTerminalService!.openTerminal(config.terminal.external, cwd);
-				});
-			}
-		});
-	}
-});
+			});
+		}
+	});
+}
+
+registerOpenTerminalCommand(OPEN_IN_INTEGRATED_TERMINAL_COMMAND_ID, 'integrated');
+registerOpenTerminalCommand(OPEN_IN_EXTERNAL_TERMINAL_COMMAND_ID, 'external');
 
 export class ExternalTerminalContribution extends Disposable implements IWorkbenchContribution {
-	private _openInTerminalMenuItem: IMenuItem;
+	private _openInIntegratedTerminalMenuItem: IMenuItem;
+	private _openInExternalTerminalMenuItem: IMenuItem;
+
+	private _IntegratedOpenEditors?: IDisposable;
+	private _IntegratedExplorer?: IDisposable;
+
+	private _ExternalOpenEditors?: IDisposable;
+	private _ExternalExplorer?: IDisposable;
 
 	constructor(
 		@IConfigurationService private readonly _configurationService: IConfigurationService
 	) {
 		super();
 
-		this._openInTerminalMenuItem = {
+		this._openInIntegratedTerminalMenuItem = {
 			group: 'navigation',
 			order: 30,
 			command: {
-				id: OPEN_IN_TERMINAL_COMMAND_ID,
+				id: OPEN_IN_INTEGRATED_TERMINAL_COMMAND_ID,
 				title: nls.localize('scopedConsoleAction', "Open in Terminal")
```

Comment:
Ideally this would say `Open in Integrated Terminal` if both are showing, not a big deal if it's too hard though.